File: go/walk.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu

func walkVals(v Val, visit func(Val) bool, seen map[Val]bool) {
	if v == nil || seen[v] {
		return
	}
	seen[v] = true

	if !visit(v) {
		return
	}

	switch n := v.(type) {
	case *MapVal:
		for _, k := range n.keys {
			walkVals(n.peg[k], visit, seen)
		}
		if n.spread != nil {
			walkVals(n.spread, visit, seen)
		}
	case *ListVal:
		for _, e := range n.peg {
			walkVals(e, visit, seen)
		}
		if n.spread != nil {
			walkVals(n.spread, visit, seen)
		}
	case *ConjunctVal:
		for _, t := range n.peg {
			walkVals(t, visit, seen)
		}
	case *DisjunctVal:
		for _, t := range n.peg {
			walkVals(t, visit, seen)
		}
	case *PrefVal:
		walkVals(n.peg, visit, seen)
		walkVals(n.superpeg, visit, seen)
	case *PlusOpVal:
		for _, t := range n.peg {
			walkVals(t, visit, seen)
		}
	case *FuncVal:
		for _, a := range n.peg {
			walkVals(a, visit, seen)
		}
	case *NilVal:
		walkVals(n.primary, visit, seen)
		walkVals(n.secondary, visit, seen)
	case *ConstraintVal:
		// A pending atom's arguments are its peg in TypeScript
		// (`new ConstraintVal({peg: args})`), so the generic walk there
		// reaches them and this one must too. The musts and the folded
		// bounds are NOT peg entries in either port.
		if n.pending != nil {
			for _, a := range n.pending.args {
				walkVals(a, visit, seen)
			}
		}
		// The value a `must` checks against is off-peg too, and it is
		// the operand a failed check reports.
		for _, m := range n.musts {
			walkVals(m.v, visit, seen)
		}
	}
}

func stampURL(v Val, url string) map[string]bool {
	urls := map[string]bool{url: true}
	walkVals(v, func(n Val) bool {
		if "" == n.srcurl() {
			n.setSrcurl(url)
		}
		urls[n.srcurl()] = true
		return true
	}, map[Val]bool{})
	return urls
}

func collectNils(v Val, out *[]*NilVal, seen map[Val]bool) {
	walkVals(v, func(n Val) bool {
		if nv, ok := n.(*NilVal); ok {
			*out = append(*out, nv)
			return false
		}
		return true
	}, seen)
}

File: ts/src/walk.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


export function walkVals(
  v: any,
  visit: (v: any) => boolean,
  seen: Set<any>
) {
  if (null == v || 'object' !== typeof v || true !== v.isVal) {
    return
  }
  if (seen.has(v)) {
    return
  }
  seen.add(v)

  if (!visit(v)) {
    return
  }

  const peg = v.peg
  if (Array.isArray(peg)) {
    for (const c of peg) {
      walkVals(c, visit, seen)
    }
  }
  else if (null != peg && 'object' === typeof peg) {
    for (const k in peg) {
      walkVals(peg[k], visit, seen)
    }
  }

  const spread = v.spread?.cj
  if (spread) {
    walkVals(spread, visit, seen)
  }

  walkVals(v.superpeg, visit, seen)
  for (const must of (v.musts ?? [])) {
    walkVals(must?.v, visit, seen)
  }

  walkVals(v.primary, visit, seen)
  walkVals(v.secondary, visit, seen)
}


export function collectNils(root: any, seen: Set<any>): any[] {
  const out: any[] = []
  walkVals(root, (v: any) => {
    if (true === v.isNil) {
      out.push(v)
      return false
    }
    return true
  }, seen)
  return out
}
